### PR TITLE
NOTIFPLT-49:  Implement a JpaNotificationService

### PR DESCRIPTION
**Pull request by drewwills**
https://issues.jasig.org/browse/NOTIFPLT-49

## See Also

* https://wiki.jasig.org/display/PLT/JpaNotificationService+-+Push-Model+Notifications

### CODE REVIEW

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/INotificationDao.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+package org.jasig.portlet.notice.service.jpa;
+
+import java.util.List;
+import java.util.Set;
+
+import org.jasig.portlet.notice.NotificationState;
+
+/* package-private */ interface INotificationDao {
```
Comment (apetro) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/INotificationDao.java]: I didn't immediately understand why the `.jpa` package would need a package-private `DAO` interface that returns `JpaEntry` objects.  JavaDoc might help articulate what this interface is intended for.

**Comment by apetro** [comment]
High level feedback:

With the caveat that I only barely skimmed, and intend to circle back and try again at understanding later.

Architecturally, this changeset appears to conflate some Notification-domain features (some new state, audit trail log, features requiring persistence) with the implementation detail of how those features are implemented (via JPA).  I imagine it would be a stronger implementation not to conflate these things.  Perhaps what is being invented here is Persistent notifications.  Or maybe what is being invented here is, well, just plain old notifications, with more of the features one would want notifications to have.  Anyway, **I wish the JPA-ness was behind an abstraction boundary from which it does not escape**, both ala [JPA Dao Guidelines](https://wiki.jasig.org/display/UPC/JPA+DAO+Guidelines) and [POJOs DAOs Registries and Services](https://wiki.jasig.org/display/UPC/POJOs+DAOs+Registries+and+Services), and ala calling the persistence-leveraging features something other than `Jpa...`.

This changeset is also making me **wish this `NotificationsPortlet` project wasn't so monolithic.**  It would be cleaner to separate the Portlet layer for accessing notifications functionality from the (potentially various) ways of implementing that functionality as well as from the API for that functionality.  I can readily imagine wanting to make use of some of the Notifications implementation to feed, say, JSON-delivering REST controllers feeding an AngularJS front end, without wanting to drag in the Portlet.  This changset didn't make this portlet project monolithic, of course, and the project deserves some credit for the existing separation of some API from implementation, but this changeset is making me think maybe this is the juncture to tease this project apart into smaller pieces.

Only skimmed; will circle back and understand more when time allows.

**Comment by drewwills** [comment]
@apetro -- Perhaps the clearest way to sum up "what's being invented" with this changeset would be:  Notifications (the project) as the system of record for notifications (the little notes that users get).  The data source abstraction in Notifications is INotificationService.  We already have several flavors of that -- e.g. RSS, RESTful-JSON over HTTP, RESTful-JSON over classpath, and Portlet Events.  All of those, as well as the JSR-286 code, are unchanged by this effort.

A JpaNotificationService is added, however.  It is one (read-only) client of the new CRUD capabilities.  The more ambitious client of those same capabilities is the bit @jhelmer-unicon is working on:  the REST API.  That interface will exercise the CRUD capabilities more comprehensively and will not speak portlet.

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaAction.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Table;
+
+/**
+ * Represents a behavior invokeable on a notification.
```
Comment (jameswennmacher) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaAction.java]: spelling: invocable

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaEvent.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+@Entity
+@Table(name=JpaNotificationService.TABLENAME_PREFIX + "EVENT")
+/* package-private */ class JpaEvent {
+
+    @Id
+    @GeneratedValue(strategy=GenerationType.AUTO)
+    @Column(name="ID", nullable = false)
+    private long id;
+
+    @OneToOne(fetch=FetchType.LAZY)
```
Comment (jameswennmacher) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaEvent.java]: Pointed out by Charles H to me; should be Many to One

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaNotificationDao.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        JpaEntry y = entityManager.merge(entry);  // Insures that cascades will be handled properly
+        entityManager.remove(y);
+    }
+
+    @Override
+    public Set<JpaEntry> getEntriesByRecipient(String username) {
+        Validate.notEmpty(username, "Argument 'username' cannot be empty");
+
+        final String jpql = "SELECT e FROM JpaEntry e WHERE e.id = ANY ("
+                + "SELECT v FROM JpaEvent v WHERE v.username = :username))";
```
Comment (jameswennmacher) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaNotificationDao.java]: Should be SELECT v.entry or v.entry.id?

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaNotificationService.java [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        List<NotificationAttribute> rslt = new ArrayList<NotificationAttribute>();
+        for (JpaAttribute a : attributes) {
+            NotificationAttribute n = new NotificationAttribute(a.getName(), a.getValues());
+            rslt.add(n);
+        }
+        return rslt;
+    }
+
+    private List<NotificationAction> prepareActions(Set<JpaAction> actions,
+            Set<NotificationState> activeStates, String username) {
```
Comment (jameswennmacher) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaNotificationService.java]: activeStates is not used.  Needed?

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaAddressee.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Represents a named addressee for a notification.  There are two types of
+ * addressee:  individual and group.  The addressees list is the target(s) of
+ * the notification as the author selected them, e.g. "send this notification to
+ * graduating seniors."  The (aggregated) recipients list is the list of all
+ * users who received the notification, e.g. "johndoe," "janedoe," and
+ * "robsmith."  A single user <em>may</em> belong to two (or more) addressee
+ * groups for a single notification, but should only see <em>one copy</em> of
+ * the notification in such a case.
```
Comment (jameswennmacher) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaAddressee.java]: Might be good to mention that adressee's are optional and used for tracking notifications sent to a group.

**Comment by jameswennmacher** [comment]
+1 recognizing it is still a work in progress, but there are needs to get this merged in as there is work building on it and I don't see anything glaring that should hold it up other than the minor fixes I mentioned.  

I'm still getting used to the idea of keeping all NotificationStates for a notification (e.g. keeping all NotificationEvents which have NotificationState and at run-time determining which states are not active).  I'm more comfortable with keeping the set of active states and have methods to add and simultaneously optionally remove a state with a transaction history of state additions/removals, though it is more data to track in the normal use cases (maybe you'd only need to track removals if you care).  However I don't want to hold up this change set. I'm thinking the current approach will handle the use cases just fine.

@apetro  I think separating out the service and dao implementation from the webapp might be reasonable and is a great observation.  I think that should be a separate change set if it is done as it affects all existing service impls.  This change set does follow the current conventions in the portlet.

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/INotificationDao.java (hunk 1 above)
Comment (drewwills) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/INotificationDao.java]: Good idea.  Incoming.

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaNotificationDao.java (hunk 4 above)
Comment (drewwills) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaNotificationDao.java]: SELECT DISTINCT v.entry FROM JpaEvent v WHERE v.username = :username

File: notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaNotificationDao.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    private EntityManager entityManager;
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Search for a JpaEntry with the specified Id. If the entry exists in the
+     * persistence context, it is returned;  otherwise, null is returned.
+     */
+    @Override
+    public JpaEntry getEntry(long entryId) {
```
Comment (chasegawa) [inline notification-portlet-webapp/src/main/java/org/jasig/portlet/notice/service/jpa/JpaNotificationDao.java]: I discovered during testing that Postgres has an issue with the fact that some of these methods do not have @Transactional on them. Without that, I would get errors like: 

Caused by: org.postgresql.util.PSQLException: Large Objects may not be used in auto-commit mode.

(which makes little sense as we are not using clobs or blobs, but that isn't the point).

File: notification-portlet-webapp/src/main/resources/META-INF/persistence.xml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            <!-- property name="hibernate.cache.provider_class" value=">net.sf.ehcache.hibernate.SingletonEhCacheProvider" /-->
+            <property name="hibernate.cache.region.factory_class" value="org.hibernate.cache.ehcache.SingletonEhCacheRegionFactory" />
+            <property name="hibernate.cache.use_minimal_puts" value="true" />
+
+            <property name="hibernate.current_session_context_class" value="thread" />
+            <!-- Ideally we'd prefer this off in production, but since the porlet (without uPortal overlay) uses in-memory
+                 hsqldb we need this to be update. Users *should* do a schema update before switching to new versions, or
+                 this should hopefully not do something negative if it ever does update the schema (typically doesn't). -->
+            <property name="hibernate.hbm2ddl.auto" value="update" />
+        </properties>
```
Comment (chasegawa) [inline notification-portlet-webapp/src/main/resources/META-INF/persistence.xml]: Need to include an additional property: 
`<property name="hibernate.dialect" value="${environment.build.hibernate.dialect}"/>`
Postgres will not work without this being set.

File: notification-portlet-webapp/src/main/resources/META-INF/persistence.xml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        <class>org.jasig.portlet.notice.service.jpa.JpaAction</class>
+        <class>org.jasig.portlet.notice.service.jpa.JpaAddressee</class>
+        <class>org.jasig.portlet.notice.service.jpa.JpaAttribute</class>
+        <class>org.jasig.portlet.notice.service.jpa.JpaEntry</class>
+        <class>org.jasig.portlet.notice.service.jpa.JpaEvent</class>
+        <properties>
+            <!-- property name="hibernate.ejb.cfgfile" value="/notification.cfg.xml" /-->
+            <property name="hibernate.session_factory_name_is_jndi" value="false" />
+            <property name="hibernate.format_sql" value="true" />
+            <property name="hibernate.generate_statistics" value="true" />
```
Comment (chasegawa) [inline notification-portlet-webapp/src/main/resources/META-INF/persistence.xml]: Adding this to make it easy to turn on better debugging would be a nice addition:
            
```
<!-- For additional detail/debugging 
            <property name="hibernate.show_sql" value="true" />
            <property name="hibernate.use_sql_comments" value="true"/>
 -->
```

File: notification-portlet-webapp/src/main/resources/datasource.properties [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+# under the License.
+#
+
+##### Default HSQL Configuration
+hibernate.connection.driver_class=org.hsqldb.jdbc.JDBCDriver
+hibernate.connection.url=jdbc:hsqldb:mem:notice
+hibernate.connection.username=sa
+hibernate.connection.password=
+hibernate.connection.validationQuery=select 1 from INFORMATION_SCHEMA.SYSTEM_USERS
+hibernate.dialect=org.hibernate.dialect.HSQLDialect
```
Comment (chasegawa) [inline notification-portlet-webapp/src/main/resources/datasource.properties]: Should be configurable:

```
hibernate.connection.driver_class=${environment.build.hibernate.connection.driver_class}
hibernate.connection.url=${environment.build.hibernate.connection.url}
hibernate.connection.username=${environment.build.hibernate.connection.username}
hibernate.connection.password=${environment.build.hibernate.connection.password}
hibernate.connection.validationQuery=${environment.build.hibernate.connection.validationQuery}
hibernate.dialect=${environment.build.hibernate.dialect}
```

Comment (drewwills) [inline notification-portlet-webapp/src/main/resources/datasource.properties]: @chasegawa -- Absolutely.  But this approach with Maven filters is something we do with _bundled_ portlets in uportal-portlets-overlay.  When we update the version pulled into uPortal, we should definitely add these.

**Pull request merged**